npc-utils.js

```diff
@@ -0,0 +1,551 @@
+
+import * as THREE from 'three';
+import {
+  rootScene,
+} from './renderer.js';
+import physicsManager from './physics-manager.js';
+
+const identityQuaternion = new THREE.Quaternion();
+
+const localVector = new THREE.Vector3();
+const localVoxel = new THREE.Object3D();
+
+const colorIdle = new THREE.Color('rgb(221,213,213)');
+const colorReached = new THREE.Color('rgb(171,163,163)');
+const colorFrontier = new THREE.Color('rgb(92,133,214)');
+const colorStart = new THREE.Color('rgb(0,255,255)');
+const colorDest = new THREE.Color('rgb(255,255,0)');
+const colorPath = new THREE.Color('rgb(149,64,191)');
+const colorPathSimplified = new THREE.Color('rgb(69,0,98)');
+
+class PathFinder {
+  constructor({voxelHeight = 1.5, heightTolerance = 0.6, detectStep = 0.1, maxIterdetect = 1000, maxIterStep = 1000, maxVoxelCacheLen = 10000, debugRender = false}) {
+    /* args:
+      voxelHeight: Voxel height ( Y axis ) for collide detection, usually equal to npc's physical capsule height. X/Z axes sizes are hard-coded 1 now.
+      heightTolerance: Used to check whether currentVoxel can go above to neighbor voxels.
+      detectStep: How height every detecting step moving.
+      maxIterdetect: How many steps can one voxel detecing iterate.
+      maxIterStep: How many A* path-finding step can one getPath() iterate. One A* step can create up to 4 voxels, 0 ~ 4.
+      maxVoxelCacheLen: How many detected voxels can be cached.
```

Comment:
Why does the cache matter? Why does the user control it? Will it change functionality to change this value?

Reply:
Every voxel need to know whether it can go to -X/+X/-Z/+Z ( left/right/btm/top) 4 neighbor voxels, based on their `position.y`.
And every voxel's `position.y` is decided by recursively `detect()`, which is sometimes a time-consuming operation.
So I cached their `position.y`, `canLeft/Right/Btm/Top` and other calculated infos for better performance.
Then I think we can't infinitely cache them all, so used `maxVoxelCacheLen` to limit the cache size.
`maxVoxelCacheLen` will not change functionality at all, users can control it to a suitable value to just prevent occupy too many memory.

But I just noticed that after refactored the algorithm from predefined two layers to support arbitrary layers, the cache seems not reused by the algorithm it self, just affecting `debugRender`.
I'll confirm and try to:
1. Reuse the cache.
2. Delete the cache and related codes at all, if performance is ok.

Reply:
Fixed "not using cache" bug. https://github.com/webaverse/app/pull/2298/commits/39b30e236fe3ab74654dcc0a654ed54ac231c243

And tested that if not using cache, the `detect()` will run much more times thus the performance will drop a lot.